### DISCUSSION TITLE
How to get all the tiles seen in the view in MapLibre

### DISCUSSION THREAD

**Discussion by dbc1129** [maplibre/maplibre-gl-js#2948] [Q&A]
How to get all the tiles seen in the view in MapLibre

**Comment by lefuturiste** [comment]
> How to get all the tiles seen in the view in MapLibre

What do you mean? Getting the raw data, or getting the list of tiles coordinates ever seen in the view ?

**Comment by HarelM** [comment]
This is complicated as when you use pitch there's a whole algorithm to calculate which tiles to use.
Especially when pitch is over 60 degrees.
You can look in the code to see the calculation, but I'm not sure you'll learn a lot from it.
In general you can take the current zoom and center and do a rough calculation of the tiles in the screen (using also the projection from pixels to lon lat), but it's tricky and not straight forward...

**Comment by usaken** [comment]
I want to know too.
How does MapLibre decide coordinates of tiles to display ?
For example, When using 256 pixels raster tile.
Which source files should I read ?

**Comment by acalcutt** [comment]
Are you able to use something simple, like the network tab of the developer tools? If you have a map with hash enabled you can zoom/pitch to a specific view, then refresh the page and see what gets loaded just for that view in the network tab
![image](https://github.com/maplibre/maplibre-gl-js/assets/3792408/5a9ef722-ab72-425d-9800-3448fba448df)

Another tool called [vt2geojson](https://github.com/mapbox/vt2geojson) was useful in seeing what was actually in those pbf files shown in the network developers tool, for example, I have used it [troubleshooting](https://github.com/maplibre/maplibre-native/issues/1843#issuecomment-1799388230)

**Comment by usaken** [comment]
Thank you for your suggestion. But I want to handle those tile coordinates in JavaScript.
I want to create a web application to plot data managed based on tile coordinates using MapLibre.

**Comment by neodescis** [comment]
I've recently needed to do something like this. I needed to get the list of vector tiles for a given source that are currently in view. This seems to do the trick for me:

```
const sourceCache = mapInstance.style.sourceCaches[sourceId];
const vectorTiles = sourceCache.getRenderableIds().map((id) => sourceCache.getTileByID(id));
```

The sourceCache also provides a `loaded()` function that will tell you if all tiles in view are loaded. This can be checked in a 'sourcedata' event handler, which you can filter by source ID. However, loaded() doesn't always seem to return true when the last 'sourcedata' event for the source fires, so I've also had to keep checking after the last 'sourcedata' event in that case (using a setTimeout or setInterval).

**Reply by usaken** [reply]
Nice idea ! thank you so much.
I think "getRenderableIds()" will make my wish come true.
And your advice for sourcedata event will help me.
Thank you for understanding my poor english.

**Reply by usaken** [reply]
I investigated the behavior of `getRenderableIds()`. This method looks returning the cached tile keys. I think that returned tile keys may contain unused tile key depending on the case. Is it wrong ?
If so, it is inconvenient for me.
I continue searching another way to get the tile coordinates in view. Any idea is welcomed.
Thank you.

**Reply by neodescis** [reply]
When I was going through the code for SourceCache, it looked like there were two basic buckets: `_tiles` and `_cache`. The `_tiles` bucket is for "current" tiles, and is what `getRenderableIds()` pulls from. The `_cache` bucket is where tiles are sent when they are no longer in the current view, but might be needed again.

Of course this is only my understanding, but it appears to hold true as I've scrolled and zoomed around the map.

**Reply by usaken** [reply]
Thank you very much.
I read source code and found my understanding was wrong.

I changed my source to read tile coordinates directly from _tiles. Due to my circumstances, I often call `clearTiles()`. `clearTiles()` seems to clear `_cache` perfectly and  abandon the data of `_tiles` but leave the tile coordinates. It is convenient for me.
When I don't use `clearTiles()`, I call `getRenderableIds()`, and when  I use, I read tile coordinates directly from `_tiles`.
